File: src/models/market_model.py
from dataclasses import dataclass
from datetime import datetime
from typing import Optional, Literal


@dataclass
class OptionContract:
    symbol: str
    strike_price: float
    expiry: datetime
    contract_id: str
    underlying: str
    is_call: bool
    current_price: float = 0.0
    volume: float = 0.0
    open_interest: float = 0.0
    
    @property
    def time_to_expiry(self) -> float:
        now = datetime.now()
        return (self.expiry - now).total_seconds() / (365 * 24 * 60 * 60)


@dataclass
class OptionAnalysis:
    contract: OptionContract
    implied_volatility: float
    theoretical_price: float
    intrinsic_value: float
    extrinsic_value: float
    greeks: dict[str, float]


@dataclass
class MarketData:
    timestamp: datetime
    price: float
    volume: float
    high: float
    low: float
    open: float
    close: float


@dataclass
class Exchange:
    name: str
    api_key: Optional[str] = None
    secret: Optional[str] = None
    testnet: bool = False

File: src/models/market.model.ts
export class Market {
    symbol: string;
    price: number;
    volume: number;

    constructor(symbol: string, price: number, volume: number) {
        this.symbol = symbol;
        this.price = price;
        this.volume = volume;
    }
}

export interface OptionContract {
    symbol: string;
    type: 'call' | 'put';
    strike: number;
    expiry: Date;
    underlying: string;
    lastPrice?: number;
    volume?: number;
    openInterest?: number;
}

export interface Greeks {
    delta: number;
    gamma: number;
    theta: number;
    vega: number;
    rho?: number;
}

export interface OptionAnalysis extends OptionContract {
    greeks: Greeks;
    impliedVolatility: number;
    theoreticalPrice: number;
    extrinsicValue: number;
    intrinsicValue: number;
}

export interface MarketData {
    timestamp: Date;
    price: number;
    volume: number;
    high: number;
    low: number;
    open: number;
    close: number;
}

export interface Exchange {
    name: string;
    apiKey?: string;
    secret?: string;
    testnet?: boolean;
}